modules/core/src/webgl/classes/texture.js

```diff
@@ -489,47 +499,53 @@ export default class Texture extends Resource {
   // Image 3D copies from Typed Array or WebGLBuffer
   setImage3D({
     level = 0,
-    internalformat = GL.RGBA,
+    dataFormat = GL.RGBA,
     width,
     height,
     depth = 1,
     border = 0,
     format,
     type = GL.UNSIGNED_BYTE,
     offset = 0,
-    pixels
+    data,
+    parameters = {}
   }) {
-    if (ArrayBuffer.isView(pixels)) {
-      this.gl.texImage3D(
-        this.target,
-        level,
-        internalformat,
-        width,
-        height,
-        depth,
-        border,
-        format,
-        type,
-        pixels
-      );
-      return this;
-    }
+    this.gl.bindTexture(this.target, this.handle);
 
-    if (pixels instanceof Buffer) {
-      this.gl.bindBuffer(GL.PIXEL_UNPACK_BUFFER, pixels.handle);
-      this.gl.texImage3D(
-        this.target,
-        level,
-        internalformat,
-        width,
-        height,
-        depth,
-        border,
-        format,
-        type,
-        offset
-      );
-    }
+    withParameters(this.gl, parameters, () => {
+      if (ArrayBuffer.isView(data)) {
+        this.gl.texImage3D(
+          this.target,
+          level,
+          dataFormat,
+          width,
+          height,
+          depth,
+          border,
+          format,
+          type,
+          data
+        );
+      }
+
+      if (data instanceof Buffer) {
```

Comment:
Side note: In some places I also support a `WebGLBuffer` being passed directly, as a "hack" on can do this:

```
data.handle || data
```

Reply:
Do you mean doing something like this?
```js
data = data.handle || data;
if (data instanceOf WebGLBuffer) {
   //...
}
```

Reply:
Yes that works.

It does raise some API design questions about if we should document that and how consistently we can apply it around the code base. And if only for buffers or also for other objects.

Right now this probably works (for buffers) in some places and not in others.